### PR TITLE
Unity ML-Agents basic tutorial

### PR DESCRIPTION
**Pull request by Dismany0**
Added a tutorial for Unity's ML-Agents Toolkit, detailing how to install and create a basic training setup where an agent learns how to choose between the larger of 2 random rewards.

Documentation and official unity sources are also linked below, along with images and gifs to assist a reader during the tutorial.

### CODE REVIEW

**Comment by Dismany0** [comment]
@ryukinouu

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+# Installation Guide
+Before installing Unity's ML-Agents package, ensure that you installed the following:
+
+[Unity 2022.3 or Later](https://unity3d.com/get-unity/download)
+
+[Python 3.10.12 or Higher](https://www.python.org/downloads/)
+
+PyTorch, it is recommended that you create a virtual environment to avoid conflicts with any existing packages.
+
+```pip3 install torch~=1.13.1 -f https://download.pytorch.org/whl/torch_stable.html```
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:16]: Specify here maybe to run the following command in your terminal

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+In this introductory tutorial, we will explore the basics of reinforcement learning with an example built completely from scratch.
+
+# Installation Guide
+Before installing Unity's ML-Agents package, ensure that you installed the following:
+
+[Unity 2022.3 or Later](https://unity3d.com/get-unity/download)
+
+[Python 3.10.12 or Higher](https://www.python.org/downloads/)
+
+PyTorch, it is recommended that you create a virtual environment to avoid conflicts with any existing packages.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:14]: Maybe you could add a quick tutorial on how to create a virtual environment here too

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+```pip3 install ./ml-agents-envs```
+
+```pip3 install ./ml-agents```
+
+Finally, within Unity's Package Manager, find and install the ```com.unity.ml-agents``` package.
+
+# What is Reinforcement Learning?
+
+In this tutorial we will use Unity's ML-Agents Toolkit to explore the basics of reinforcement learning. Reinforcement Learning is a form of machine learning where an agent will learn to make decisions by interacting with an environment. Based on the actions it performs, it will receive a reward that indicates how well it performed. Over time, the goal of the agent is to learn a strategy that maximizes the reward it receives.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:28]: Cool! But what can you use this for?

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+In order to do so, our agent will follow a simple cycle of 3 steps. It will first collect an observation from the environment. Using this information collected, it will perfmr an action. Finally, based on this action, we will give the agent a reward.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:30]: Perform*

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+In my setup, I have currently created a floor, an Agent, and 2 rewards.
+
+They are positioned along a straight line. Our environment will involve the agent being able to perform 3 actions, moving left, moving right, or doing nothing at all. Hopefully, we will be able to train this agent to move towards the larger value.
+
+Finally, to finish this section, create a C# script and attach it to your agent object. I have named the script ```BiggerRewardAgent.cs```, but feel free to change the name to anything you'd like.
+
+# Actions
+
+Now that our environment is set up, we will assign actions that our agent is able to perform. Begin by opening the ```CollectRewardAgent.cs``` script with any text editor you would like, such as Visual Studio or Visual Studio Code. We will begin by changing our MonoBehaviour into a Agent class. Make the following imports:
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:49]: What are these imports for? You could specify

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+```
+using UnityEngine;
+using Unity.MLAgents;
+using Unity.MLAgents.Sensors;
+using Unity.MLAgents.Actuators;
+```
+
+We will also removing the ```Update()``` and ```Start()``` functions, and add the following overrides:
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:58]: We will also be*

Comment (ryukinouu) [reply Topics/Tech_Stacks/Unity_ML_Agents.md:58]: and adding*

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+```public override void OnActionReceived(ActionBuffers actions)```
+
+```public void OnTriggerEnter(Collider other)```
+
+In the following few sections, we will go through and explain each of these functions in detail. We will first begin with ```OnActionReceieved()```
+
+Unlike us, our agent is only able to play through making actions through numbers. These numbers will be given through the ActionBuffer. As specified before, we want our agent to be able to move left, right, and stand still. We will accomplish this by assigning each action a number. We begin by obtaining the action from DiscreteActions, or integer actions. We will set up where our AI gets these DiscreteActions later in our observations section.
+
+```
+public override void OnActionReceived(ActionBuffers actions)
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:73]: You could add a comment or two here for people to understand what's happening

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+```
+public class CollectRewardAgent : Agent
+{
+    public GameObject rewardSmall;
+    public GameObject rewardLarge;
+    ...
+```
+
+This will allow us to easily access the position of both reward objects as well. To add an observation, we simply go under the CollectObservations function and call ```Sensor.AddObservation()```. We would like to add the x value of all 3 positions, as our agent will only be capable of moving along the x axis.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:107]: Lower case sensor in ```Sensor.AddObservation()```

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+```
+public override void CollectObservations(VectorSensor sensor)
+    {
+        sensor.AddObservation(rewardSmall.transform.localPosition.x);
+        sensor.AddObservation(rewardLarge.transform.localPosition.x);
+        sensor.AddObservation(transform.localPosition.x);
+    }
+```
+
+Here we are adding the local positions of all 3 objects. I use local position here so that if we move our training setup around, everything stays functional. However, you can adjust this depending on how your training setup may look like.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:118]: First time you used "I" instead of "we", you should change it to keep it consistent

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    }
+```
+
+From anywhere in Unity, we are able to assign rewards to our agent using the functions ```SetReward()``` and ```AddReward()```. For our purposes, we will set a reward when the agent collides with another object, in this case, either the small or large reward. While the amount you assign to each reward is arbitrary, they should be similar relative to each other. If the rewards are too close to each other, the agent might not learn effectively. Play with your values to find an optimal value.
+
+Something you might have noticed is the ```EndEpisode()``` after we assign the reward. This means that we want our iteration of training to stop when our agent gets its reward, so that it can start training all over again. In order to do so, we have one last function we should call.
+
+```
+public override void OnEpisodeBegin()
+    {
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:146]: Same here above, could add a comment maybe

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+<img src="https://raw.githubusercontent.com/Dismany0/misc_images/main/CSC301%20Assignment%202/Screenshot%202023-11-26%20025032.png" height="500"/>
+
+Our **Vector Observation** should determine the amount of observations our agent is able to make. This should match the amount of observations we gave it in our prior example, which is 3. If your training set up differs, make sure to assign the number of observations correctly.
+
+Next, under the **Actions** section, we want to set the number of Discrete Branches to 1. This is where we are making our observations. Next, we want to set the size of our branch to 3. This should correspond to the number of possible actions our agent is able to make. As our agent is only able to move left, right, and stay still, the size of our action space is 3, as we assigned in the Actions section above.
+
+<img src="https://raw.githubusercontent.com/Dismany0/misc_images/main/CSC301%20Assignment%202/Screenshot%202023-11-26%20031918.png" height="500"/>
+
+The final step is to add a Decision Requester. The only parameter here is how often our agent will make an action. Feel free to play with this value.
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:177]: What's a decision requester?

File: Topics/Tech_Stacks/Unity_ML_Agents.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+<img src="https://github.com/Dismany0/misc_images/blob/main/CSC301%20Assignment%202/ezgif.com-video-to-gif%20(3).gif?raw=true" height="500"/>
+
+You've just learned how to train your first agent in Unity! While the agent might not appear very smart at first, it should eventually learn that moving towards the larger reward is much more rewarding than the smaller one. Over time, you should expect to see the average reward, provided every few minutes in the terminal, to increase.
+
+Feel free to make your own adjustments or create your own training environment. Fun additions you could try to build off of this example can include
+- Movement along a 2d square instead of a line
+- A penalty for taking a long time to reach a reward
+- Additional obstacles your agent must avoid
+
+Feel free to explore both on your own, and with Unity's [official documentation](https://unity-technologies.github.io/ml-agents/ML-Agents-Toolkit-Documentation/) and the [ML-Agents Repo](https://github.com/Unity-Technologies/ml-agents).
```
Comment (ryukinouu) [inline Topics/Tech_Stacks/Unity_ML_Agents.md:199]: You said feel free 6 times in the page, could look for synonyms to replace some of them

**Review by ryukinouu — CHANGES_REQUESTED** [review]
Your wiki page is really well done! The use of images and explanations add greatly to the readability of the page. I also liked how you gave the readers options on a lot of things that they could play around with.
Just requesting some minor grammatical changes and suggestions in a few places that might be a little confusing to readers. Otherwise, good job!👍

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 2 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md]: Added a small section about virtual environments and to run everything in a terminal opened to the project directory

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 3 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md]: I expanded on the introduction to also talk about the possible uses of machine learning, especially in a unity environment. I feel adding it here kinda breaks the flow

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 9 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md]: I originally used I to differentiate how this is something that I did for the project, but was mostly optional. 
To avoid confusion, I replaced it with we and reworded my explanation for why I used local positioning

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 12 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md]: haha I realized that too reading through mine another time. I replaced most of them with other phrases to make it a little less jarring to read over and over again. I also reworded the final section with the two links.

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 7 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md:78]: Added another small section explaining how the OnActionReceived function works

File: Topics/Tech_Stacks/Unity_ML_Agents.md (hunk 10 above)
Comment (Dismany0) [reply Topics/Tech_Stacks/Unity_ML_Agents.md:151]: Added section explaining episodes and how the OnEpisodeBegin function works

**Review by ryukinouu — APPROVED** [review]
Looks good!
